packages/react-shopify-app-bridge/src/Provider.tsx

```diff
@@ -84,30 +79,42 @@ const InnerGadgetProvider = memo(
       console.debug("[gadget-rsab] set up client auth for session tokens");
     }, [api.connection, appBridge]);
 
-    let runningShopifyAuth = false;
+    let redirectToOauth = false;
     let isAuthenticated = false;
+    const hasFetchedOrInstalledShop = useRef(false);
 
-    // always run one session fetch to the gadget backend on boot to discover if this app is installed
-    const [{ data: currentSessionData, fetching: sessionFetching, error }] = useQuery({
-      query: GetCurrentSessionQuery,
-    });
-
-    if (currentSessionData) {
-      runningShopifyAuth = currentSessionData.shopifyConnection?.requiresReauthentication;
+    const [{ data: fetchOrInstallShopData, fetching: fetchingOrInstallingShop, error }, fetchOrInstallShop] =
+      useMutation(FetchOrInstallShopMutation);
 
-      if (currentSessionData.currentSession) {
-        if (!currentSessionData.currentSession.shop) {
-          runningShopifyAuth = true;
-        } else {
-          // we may be missing scopes, if so, we aren't fully authenticated
-          isAuthenticated = !currentSessionData.shopifyConnection?.requiresReauthentication;
-        }
-      }
+    if (fetchOrInstallShopData) {
+      console.debug({ fetchOrInstallShopData }, "[gadget-rsab] fetched or installed shop data");
+      redirectToOauth = fetchOrInstallShopData.shopifyFetchOrInstallShop.redirectToOauth;
+      isAuthenticated = fetchOrInstallShopData.shopifyFetchOrInstallShop.isAuthenticated;
     }
 
+    // always run one fetch to the gadget backend on boot to discover if this app is installed
+    useEffect(() => {
+      if (hasFetchedOrInstalledShop.current) return;
+      if (!appBridge) return;
+
+      hasFetchedOrInstalledShop.current = true;
+
+      appBridge
+        .idToken()
+        .then((shopifySessionToken) => {
+          console.debug("[gadget-rsab] fetching or installing shop");
+          fetchOrInstallShop({ shopifySessionToken }).catch((err) => {
+            console.error({ err }, "[gadget-rsab] error fetching or installing shop");
+          });
+        })
+        .catch(() => {
+          console.error("[gadget-rsab] failed to get id token from app bridge");
```

Comment:
Can we return this error to the caller somehow instead of just swallowing it? Without that, someone can't show a toast or something if there's a network connectivity error or a 500 or whatever (or even know that it happened)

Reply:
ok made a change to support this; lmk what you think of the approach (I'm not in love with it)